design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,148 @@
+# Rook-Ceph Bucket Provisioning
+
+## Proposed Feature
+
+Add an CephObjectBucketClaim API endpoint and control loop to the Rook-Ceph operator for the dynamic provisioning of S3 buckets by Kubernetes users.
+
+### Overview
+
+  Rook-Ceph has developed Custom Resource Definitions to support the dynamic provisioning of Ceph Object Stores and Ceph Object Store Users.  This proposal introduces the next logical step of dynamic bucket provisioning with the addition of a CephObjectBucketClaim CRD.  The intent is to round out the Rook-Ceph experience for Kubernetes users by providing a single control plane for the managing of Ceph Object components.      
+
+## Goals
+
+- Provide Rook-Ceph users the ability to dynamically provision object buckets via the Kubernetes API with a CephObjectBucketClaim Custom Resource Definition (CRD).
+- Utilize a familiar and deterministic pattern when injecting bucket connection information into workload environments. 
+
+## Non-Goals
+
+- This proposal does not implement a generalized method for bucket provisioning; it will be specific to Rook-Ceph
+- This design does not attempt to implement a 1:1 model of Kubernetes PersistentVolumes and PersistentVolumeClaims.  Some shallow similarities will exist. 
+- This design does not include the Swift interface implementation by Ceph Object.
+- This design does not provide users a means of deleting buckets via the Kubernetes API.  This creates a dangerous situation which could lead to accidental loss of data.  Only the Kubernetes API objects are deleted.
+
+## Users
+
+- Admin: A Kubernetes cluster administrator with permissions to instantiate and control access to cluster storage. The admin will use existing Rook-Ceph CRDs to create object stores and StorageClasses to make them accessible to users.  
+- User: A Kubernetes cluster user with limited permissions, typically confined CRUD operations on Kubernetes objects within a single namespace.  The user will create request ceph object buckets by instantiating CephObjectBucketClaims. 
+
+## Use Cases
+
+### Assumptions
+
+1. A Rook-Ceph operator deployed on a Kubernetes Cluster.
+1. A CephObjectStore object exists along with the accompanying Service in a protected namespace.
+1. An existing CephObjectStoreUser exists along with the accompanying Secret in the user's namespace.
```

Comment:
Do we require the ceph object user to exist or just the secret. I realize in most cases a user will be created which instantiates the secret, but brownfield use cases may not need the user object.

Reply:
This raises the bigger question of "Do we support brownfield?"  The design specifies Rook-Ceph.  I would said that we couple the COBC to the CephObjectUser to further cement this point.

Reply:
If we do not want to deal with brownfield (which is fine for now) then this should be listed as an assumption.

Reply:
I think there's also a concern for separating the owner keys from the user keys.  Meaning that the operator would have bucket create/delete perms and only generate on-demand user keys with object put/delete within that bucket.  This model would complicate the brownfield case even more.  How do you know the requester should be given access to the bucket?